perf_bench_test.go

```diff
@@ -24,127 +24,240 @@ package main
+	require.Equal(t, resp.IsColdStart, false)
+	if err != nil {
+		log.Warn("Function returned error")
 	} else {
-		require.Equal(t, resp.IsColdStart, false)
+		atomic.AddInt64(realRPS, 1)
+	}
+	if resp.Payload != "Hello, replay_response!" {
+		log.Warnf("Function returned invalid: %s", resp.Payload)
+	}
+
+	execTime := time.Since(tStart).Milliseconds()
+	atomic.AddInt64(avgExecTime, execTime)
+	log.Debugf("VM %s: returned in %d milliseconds", vmIDString, execTime)
+}
+
+// Returns a list of image names
+func getImages(t *testing.T) []string {
+	var (
+		images = map[string]string{
+			"helloworld":   "ustiugov/helloworld:var_workload",
+			"chameleon":    "ustiugov/chameleon:var_workload",
+			"pyaes":        "ustiugov/pyaes:var_workload",
+			"image_rotate": "ustiugov/image_rotate:var_workload",
+			"json_serdes":  "ustiugov/json_serdes:var_workload",
+			"lr_serving":   "ustiugov/lr_serving:var_workload",
+			"cnn_serving":  "ustiugov/cnn_serving:var_workload",
+			"rnn_serving":  "ustiugov/rnn_serving:var_workload",
+		}
+		funcs  = strings.Split(*funcNames, ",")
+		result []string
+	)
+
+	for _, funcName := range funcs {
+		imageName, isPresent := images[funcName]
+		require.True(t, isPresent, "Function %s is not supported", funcName)
+		result = append(result, imageName)
+	}
+
+	return result
+}
+
+func writeResultToCSV(t *testing.T, outfile string, metrics map[string]float64) {
+	outFile := getOutFile(outfile)
+
+	f, err := os.OpenFile(outFile, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0644)
+	require.NoError(t, err, "Failed opening file")
+	defer f.Close()
+
+	reader := csv.NewReader(f)
+	headers, err := reader.Read()
+	require.True(t, err == nil || err.Error() == "EOF", "Failed reading file")
+
+	writer := csv.NewWriter(f)
+	if headers == nil {
+		for k := range metrics {
+			headers = append(headers, k)
+		}
+		err = writer.Write(headers)
+		require.NoError(t, err, "Failed writting file")
+		writer.Flush()
+	}
+
+	var data []string
+	for _, header := range headers {
+		for k, v := range metrics {
+			if k == header {
+				vStr := strconv.FormatFloat(v, 'f', -1, 64)
+				data = append(data, vStr)
+			}
+		}
 	}
+	err = writer.Write(data)
+	require.NoError(t, err, "Failed writting file")
 
-	log.Debugf("vmID %s: returned in %f seconds", vmIDString, time.Since(start).Seconds())
+	writer.Flush()
 }
 
-func getImages() map[string]string {
-	return map[string]string{
-		"helloworld": "ustiugov/helloworld:var_workload",
-		// "chameleon":    "ustiugov/chameleon:var_workload",
-		// "pyaes":        "ustiugov/pyaes:var_workload",
-		// "image_rotate": "ustiugov/image_rotate:var_workload",
-		// "json_serdes":  "ustiugov/json_serdes:var_workload",
-		// "lr_serving":   "ustiugov/lr_serving:var_workload",
-		// "cnn_serving":  "ustiugov/cnn_serving:var_workload",
-		// "rnn_serving":  "ustiugov/rnn_serving:var_workload",
-		// "lr_training":  "ustiugov/lr_training:var_workload",
+func getCPUIntenseRPS() map[string]int {
```

Comment:
please add a description of this function in a comment

Reply:
Done